src/client/datascience/jupyter/kernelVariables.ts

```diff
@@ -471,22 +471,50 @@ export class KernelVariables implements IJupyterVariables {
         return result;
     }
 
-    private async getVariableNamesFromKernel(notebook: INotebook, token?: CancellationToken): Promise<string[]> {
+    private async getVariableNamesAndTypesFromKernel(
+        notebook: INotebook,
+        token?: CancellationToken
+    ): Promise<IJupyterVariable[]> {
         // Get our query and parser
         const query = this.getParser(notebook);
 
         // Now execute the query
         if (notebook && query) {
+            // Add in our get variable info script to get types
+            await this.importGetVariableInfoScripts(notebook, token);
+
             const cells = await notebook.execute(query.query, Identifiers.EmptyFileName, 0, uuid(), token, true);
             const text = this.extractJupyterResultText(cells);
-
-            // Apply the expression to it
             const matches = this.getAllMatches(query.parser, text);
+            const matchesAsStr = matches.map((v) => `'${v}'`);
+
+            // VariableTypesFunc takes in list of vars and the corresponding var names
+            const results = await notebook.execute(
+                `print(${GetVariableInfo.VariableTypesFunc}([${matches}], [${matchesAsStr}]))`,
```

Comment:
This is obviously python specific. The query.query was intended to support other languages. I think we could at least extract out the python part into a separate class? Maybe make an interface for getting names and types, like IKernelVariableRequester or something.

So instead of having getVariableNamesandTypesFromKernel be a function, there would be an interface with this single method on it. 

Right now we'd just support python, but this interface could be expanded (or suppled by other extensions) for other languages later.

Reply:
There are other functions that do some python specific things. I'm thinking about putting those into the PythonVariableRequester as well.